consensus/src/lib.rs

```diff
@@ -330,6 +343,60 @@ impl<TYPES: NodeType, LEAF: LeafType<NodeType = TYPES>> Consensus<TYPES, LEAF> {
         let leaf = view
             .get_leaf_commitment()
             .expect("Decided state not found! Consensus internally inconsistent");
-        self.saved_leaves.get(leaf).unwrap().clone()
+        self.saved_leaves.get(&leaf).unwrap().clone()
+    }
+}
+
+/// Mapping from block commitments to full blocks.
+///
+/// Entries in this mapping are reference-counted, so multiple consensus objects can refer to the
+/// same block, and the block will only be deleted after _all_ such objects are garbage collected.
+/// For example, multiple leaves may temporarily reference the same block on different branches,
+/// before all but one branch are ultimately garbage collected.
+#[derive(Clone, Debug, Derivative)]
+#[derivative(Default(bound = ""))]
+pub struct BlockStore<Block: Committable>(HashMap<Commitment<Block>, (Block, u64)>);
```

Comment:
nit: we've been following the convention of screaming snake case for generics (to make it obvious that they're generics). For consistency's sake it might make sense to do:
```suggestion
pub struct BlockStore<BLOCK: Committable>(HashMap<Commitment<BLOCK>, (BLOCK, u64)>);
```